doc/guide/dts: DTS documentation add-on

erwango opened the pull request:
Update dts documentation.
Rework some parts to provide more clarity and elaborate
yaml bindings usage.

Fixes #10318

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -16,18 +16,21 @@ PowerPC architecture.  However, it is now in use for ARM and other
+resources that are the primary target of Zephyr.
+Hence, in Zephyr, the DTS information will be used only during compile time.
 Information about the system is extracted from the compiled DTS and used to
 create the application image.
```
When I read this it sounds like two paragraphs of information that then abruptly ends with; "but we don't use this, so none of what you read is relevant."

Can we cut down on how much we talk about DTB. Since DTB is not in use, one would think we would be able to document our DT usage without even mentioning it.

ulfalizer replied:
FWIW, I like the DTB stuff. Gives some context when looking up device tree info elsewhere.

erwango replied:
> Can we cut down on how much we talk about DTB?

I mostly agree, though it could be interesting at least explaining why we're not using dtb the usual way and process dts at target compile time. This is not obvious to every one, I already had to explain this a couple of times to people used to Linux and that can't figure out the memory constraints we're facing.
So this is more targeting Linux users than non Linux aware developers, but still a useful piece of information. Anyway, we can try to shorten it a bit.

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -169,8 +172,7 @@ of things will vary based on the board layout and application use.
 
-Device tree is currently supported on all ARM targets.  Support for all other
-architectures is to be completed by release 1.11.
+Device tree is currently supported on all targets except posix (boards/posix).
```
xtensa?

Is this comment outdated?

https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/dts.cmake#L26

SebastianBoe replied:
Also, a reader might not be familiar with boarsd/posix.

So maybe instead say "Device tree is currently supported on all embedded targets".

erwango replied:
I found dts files in all extensa boards...

erwango replied:
> So maybe instead say "Device tree is currently supported on all embedded targets"

Well, it is supported by some qemu variants.
Sticking with posix and adding qemu_x86_64 whifch I initially forgot.

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -341,10 +347,10 @@ hierarchy looks like the following::
-architectures.
+The common directory contains a skeleton.dtsi which provides device tree root
+node definition.  The bindings subdirectory contains YAML binding files used
+to instruction how python DTS parsing script should extract nodes information
```
Something is wrong with the english:

"used to instruction how"

erwango replied:
Was supposed to be ""used to instruct how"". Txs

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -356,25 +362,31 @@ Example: Subset of DTS/YAML files for NXP FRDM K64F (Subject to Change)::
+describe the contents or definition of a device tree node and instruct how the
+extracted information should be formatted.
+
+A YAML description (a.k.a YAML binding) must be provided for every device node
```
Do you need to introduce the term "YAML description" ?

Can we stick to "YAML binding" ? Or is there a difference?

erwango replied:
Actually YAML description is what it is, YAML bindings is how we call them.
Indeed, "a.k.a" makes it unclear. Let me reword it.

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -356,25 +362,31 @@ Example: Subset of DTS/YAML files for NXP FRDM K64F (Subject to Change)::
+is associated to each node `compatible` property.  Information within the YAML
+file will instruct the python dts parsing script (located in `scripts/dts`) how
+each property of the node is expected to be generated, either the type of the
+value or the format of it's name.  For now, node properties are generated as
```
Drop "For now," all documentation is "For now".

And being able to access node properties through defines is not likely to go away any time soon.

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -356,25 +362,31 @@ Example: Subset of DTS/YAML files for NXP FRDM K64F (Subject to Change)::
+In a next step, YAML bindings could  also be used to validate the device tree
+files themselves.  A device tree file can successfully compile and still not
+contain the necessary pieces required to build the rest of the software.  YAML
+binding files provide a means to detect that issue in the future.
```
Drop this paragraph.

We should not be documenting what we would like the system to do. This is more appropriate in issues etc.

The user is reading this to understand what it does do.

SebastianBoe commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +395,39 @@ information extraction phase and are matched to device tree nodes via the
+Additionally, a concatenated file named generated_dts_board_fixups.h is
+generated in the same directory using all board related fixup files.
+
+These both files are then combined into a generated_dts_board.conf file.
```
> These two files are included by both files are then combined into a generated_dts_board.conf file.

No it's not.

.conf corresponds to generated_dts_board_unfixed.h.

You could however say that:

"The source file generated_dts_board.h should be included by C source files to access DT information as it in turn includes generated_dts_board_unfixed.h and generated_dts_board_fixups.h.

erwango replied:
My bad. Txs for proposal.

erwango commented:
recheck

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +390,40 @@ information extraction phase and are matched to device tree nodes via the
+Include files generation
+************************
+
+At build, after board .dts file has been processed by dtc (device tree compiler),
```
```suggestion
At build time, after a board's ``.dts`` file has been processed by the DTC (Device Tree Compiler),
```

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +390,40 @@ information extraction phase and are matched to device tree nodes via the
+************************
+
+At build, after board .dts file has been processed by dtc (device tree compiler),
+a frdm_k64f.dts_compiled file is generated under the zephyr directory.
```
```suggestion
a corresponding ``.dts_compiled`` file is generated under the ``zephyr`` directory.
```

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +390,40 @@ information extraction phase and are matched to device tree nodes via the
+
+At build, after board .dts file has been processed by dtc (device tree compiler),
+a frdm_k64f.dts_compiled file is generated under the zephyr directory.
+This file is parsed by python dts parsing script, and thanks to YAML bindings it
```
How about:
```
This ``.dts_compiled`` file is processed by the python DTS parsing script
and generates an include file named
``include/generated/generated_dts_board_unfixed.h``
that holds all the information extracted from the DTS file with
the format specified by the YAML bindings.  For example:
```

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +390,40 @@ information extraction phase and are matched to device tree nodes via the
+   #define SW1_GPIO_PIN				DT_GPIO_KEYS_BUTTON_0_GPIO_PIN
+   #define SW1_LABEL				DT_GPIO_KEYS_BUTTON_0_LABEL
+
+Additionally, a concatenated file named generated_dts_board_fixups.h is
```
```suggestion
Additionally, a file named ``generated_dts_board_fixups.h`` is
```

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +390,40 @@ information extraction phase and are matched to device tree nodes via the
+   #define SW1_LABEL				DT_GPIO_KEYS_BUTTON_0_LABEL
+
+Additionally, a concatenated file named generated_dts_board_fixups.h is
+generated in the same directory using all board related fixup files.
```
```suggestion
generated in the same directory concatenating all board-related fixup files.
```

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +390,40 @@ information extraction phase and are matched to device tree nodes via the
+Additionally, a concatenated file named generated_dts_board_fixups.h is
+generated in the same directory using all board related fixup files.
+
+The include file include/generated_dts_board.h includes both these generated
```
```suggestion
The include file ``include/generated_dts_board.h`` includes both these generated
```

dbkinder commented on doc/guides/dts/index.rst:
```diff
@@ -383,5 +390,40 @@ information extraction phase and are matched to device tree nodes via the
+generated in the same directory using all board related fixup files.
+
+The include file include/generated_dts_board.h includes both these generated
+files and hence gives access to DT information to Zephyr C source files.
```
```suggestion
files, giving Zephyr C source files access to the board's device tree information.
```

erwango commented on doc/guides/dts/index.rst:
```diff
@@ -356,25 +362,26 @@ Example: Subset of DTS/YAML files for NXP FRDM K64F (Subject to Change)::
+
+A YAML description (called "YAML binding") must be provided for every device node
+that is to be a source of information for the system.  A YAML binding file
+is associated to each node `compatible` property.  Information within the YAML
```
@dbkinder , 'compatible' is a property of a node, such as 'label' is another possiebl property.
I don't understand what a node-compatible thing would be.

erwango commented on doc/guides/dts/index.rst:
```diff
@@ -16,18 +16,20 @@ PowerPC architecture.  However, it is now in use for ARM and other
-Traditional usage of device tree involves storing of the Device Tree Blob.
-The DTB is then used during runtime for configuration of device drivers.  In
-Zephyr, the DTS information will be used only during compile time.
+Storing the DTB and the code required to parse it requires an amount of memory
```
Actually, the first proposed sentence is somewhat unclear to the non native speaker I am. What about : 
```
For larger systems, the benefit of the offered flexibility overcomes the requested footprint overhead. [..]
```

nashif replied:
i like what @dbkinder better ;)

erwango replied:
> i like what @dbkinder better ;)

Ok, let's go with that then!

erwango committed:
doc/guide/dts: DTS documentation add-on

Update dts documentation.
Rework some parts to provide more clarity and elaborate
yaml bindings usage.

Fixes #10318

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

dbkinder reviewed (approved):
LGTM, thanks.
